tests/test_parser.py

```diff
@@ -131,3 +134,121 @@ def test_string_utf8():
     parser.consume_byte(b[:1])
     assert parser.valid_next_bytes() == set([Byte(b[1:])])
     parser.consume_byte(b[1:])
+
+
+class TestRecursiveNullableGrammars:
+    """
+    Computing parse tree of recursive nullable grammars will cause an infinite
+    loop if not handled correctly
+    """
+
+    @pytest.mark.timeout(5)
+    def test_no_infinite_loop(self):
+        """
+        A -> A
+        A ->
+
+        Loop occurs because `A -> A` is a nullable rule
+        """
+        # Note that we get a different grammar if we made `A = select([''], recurse=True)`
+        A = Select([], recursive=True)
+        A.values = [A, ""]
+        parser = EarleyCommitParser(A)
+        # Test that computing the parse tree doesn't hang
+        parser.parse_tree()
+        # Test that getting captures doesn't hang
+        parser.get_captures()
+
+    @pytest.mark.timeout(5)
+    def test_no_infinite_loop_with_terminal(self):
+        """
+        A -> A B
+        A ->
+        B -> 'x'
+        B ->
+
+        Loop occurs because `A -> A B` is a nullable rule
+        """
+        B = select(["x", ""])
+        A = select([B, ""], recurse=True)
+        parser = EarleyCommitParser(A)
+        # Test that computing the parse tree doesn't hang
+        parser.parse_tree()
+        # Test that getting captures doesn't hang
+        parser.get_captures()
```

Comment:
For all of these tests, it would be nice if there was _something_ to `assert` about the state of the parser at the end.

Reply:
Agreed. Will keep thinking about this!

Reply:
Working on making some assertions here, I became dissatisfied with the fact that my fix doesn't actually prevent infinite parse trees like `A -> A -> ...`; it only prevents looping over these parse trees forever. The _behavior_ is right in that we're still able to correctly get captures, so maybe this is fine. But marking this as a draft in the meantime while I see if I can get a bit closer to the root of the issue.